pkg/controller/controller_instance.go

```diff
@@ -595,24 +606,44 @@ func (c *controller) reconcileServiceInstance(instance *v1alpha1.ServiceInstance
 			successProvisionReason,
 			successProvisionMessage,
 		)
-		c.updateServiceInstanceStatus(toUpdate)
+		c.setServiceInstanceForCompletedChange(toUpdate)
 
 		c.recorder.Eventf(instance, api.EventTypeNormal, successProvisionReason, successProvisionMessage)
 	}
-	return nil
+	return true, nil
 }
 
-func (c *controller) pollServiceInstanceInternal(instance *v1alpha1.ServiceInstance) error {
+func (c *controller) pollServiceInstanceInternal(instance *v1alpha1.ServiceInstance) (bool, error) {
 	glog.V(4).Infof("Processing ServiceInstance %v/%v", instance.Namespace, instance.Name)
 
-	serviceClass, servicePlan, brokerName, brokerClient, err := c.getServiceClassPlanAndServiceBroker(instance)
+	clone, err := api.Scheme.DeepCopy(instance)
 	if err != nil {
-		return err
+		return false, err
+	}
+	toUpdate := clone.(*v1alpha1.ServiceInstance)
+
+	serviceClass, servicePlan, brokerName, brokerClient, err := c.getServiceClassPlanAndServiceBroker(toUpdate)
+	if err != nil {
+		c.updateServiceInstanceStatus(toUpdate)
+		return false, err
+	}
+
+	// If the reconciler has exceeded the time allowed to retry failed
+	// reconciliations on the ServiceInstance, then do not reconcile it.
+	if instance.Status.OperationStartTime != nil && !time.Now().Before(instance.Status.OperationStartTime.Time.Add(c.reconciliationRetryDuration)) {
+		s := fmt.Sprintf("Stopping reconciliation retries on ServiceInstance '%v/%v' because too much time has elapsed", instance.Namespace, instance.Name)
+		glog.Info(s)
+		c.recorder.Event(instance, api.EventTypeWarning, errorReconciliationRetryTimeoutReason, s)
+		setServiceInstanceCondition(toUpdate,
+			v1alpha1.ServiceInstanceConditionFailed,
+			v1alpha1.ConditionTrue,
+			errorReconciliationRetryTimeoutReason,
+			s)
+		c.setServiceInstanceForCompletedChange(toUpdate)
+		c.updateServiceInstanceStatus(toUpdate)
+		return true, errors.New(s)
```

Comment:
Before returning, we should call `c.finishPollingServiceInstance(toUpdate)`. Otherwise, should the instance be recreated, its polling rate will continue from whatever slowed-down rate it was last at (presumably very slow, due to having continuously decayed for one week).

Reply:
Good catch. I think with the change to return the `complete` bool from pollServiceInstance, we don't actually need the finishPollingServiceInstance function any more. We can just use the return value in worker function to determine whether to requeue or not, if we were to change the worker to forget after success.

Reply:
@staebler We still will need the function, as the worker thread we're returning the `complete` boolean to is doing work for a key it got from the instance queue, but we need both the instance queue and the polling queue to forget the key. The flow of an async operation is as follows:

- When a poll starts, add the instance rate limited to the polling queue.
- When the polling queue gets to the instance's key, add it to the instance queue.
- When the instance queue gets to the instance's key, do the reconciliation (aka. call `pollServiceInstance`)
- If we need to continue polling, add rate limited to the polling queue.
- When the async operation is complete, forget the item from both the instance queue and polling queue

Reply:
(PS. Just deleted my previous follow-up comment, I think I misunderstood your recommendation)

The polling queue's reconciliation function just adds the instance to the instance queue, so there'd be no way to signal from it whether the work is complete or not. That signaling has to happen from the reconciliation function of the instance queue, which has to do the double duty of forgetting the instance from both the instance and polling queues on finishing an async operation.